charms/kfp-persistence/src/charm.py

```diff
@@ -9,191 +9,71 @@
-        try:
-            self._check_leader()
-            interfaces = self._get_interfaces()
-            image_details = self.image.fetch()
-            kfpapi = self._get_kfpapi(interfaces)
-        except (CheckFailedError, OCIImageResourceError) as check_failed:
-            self.model.unit.status = check_failed.status
-            self.log.info(str(check_failed.status))
-            return
-
-        self.model.unit.status = MaintenanceStatus("Setting pod spec")
-
-        self.model.pod.set_spec(
-            {
-                "version": 3,
-                "serviceAccount": {
-                    "roles": [
-                        {
-                            "global": True,
-                            "rules": [
-                                {
-                                    "apiGroups": ["argoproj.io"],
-                                    "resources": ["workflows"],
-                                    "verbs": ["get", "list", "watch"],
-                                },
-                                {
-                                    "apiGroups": ["kubeflow.org"],
-                                    "resources": ["scheduledworkflows"],
-                                    "verbs": ["get", "list", "watch"],
-                                },
-                                {
-                                    "apiGroups": [""],
-                                    "resources": ["namespaces"],
-                                    "verbs": ["get"],
-                                },
-                            ],
-                        }
-                    ]
+    """Charm for the data persistence application of Kubeflow Pipelines."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize charm and setup the container."""
+        super().__init__(*args, **kwargs)
+
+        # Charm logic
+        self.charm_executor = CharmReconciler(self)
+
+        # Components
+        self.leadership_gate = self.charm_executor.add(
+            component=LeadershipGateComponent(charm=self, name="leadership-gate"), depends_on=[]
+        )
+
+        self.kfp_api_relation = self.charm_executor.add(
+            component=SdiRelation(charm=self, name="relation:kfp-api", relation_name="kfp-api"),
+            depends_on=[self.leadership_gate],
+        )
+
+        self.persistenceagent_container = self.charm_executor.add(
+            component=PersistenceAgentContainer(
+                charm=self,
+                name="container:persistenceagent",
+                container_name="persistenceagent",
+                service_name="persistenceagent",
+                files_to_push=[],
+                environment={
+                    "KUBEFLOW_USERID_HEADER": "kubeflow-userid",
+                    "KUBEFLOW_USERID_PREFIX": "",
+                    # Upstream defines this in the configmap persistenceagent-config-*
+                    "MULTIUSER": "true",
```

Comment:
This is interesting.  My first instinct was to put `environment`'s data in the `PersistenceAgentContainer.get_layer()` function, but I find this neat too.  I'm curious what your thought was and what others feel on this too

Reply:
Idea was that this could be a part of charm config or anyother config that is static available at the time of creation of that compoent. All the dynamic data (i.e. from relation or library) is handled via call backs. That data still can set and/or environment variables.
Basically, trying to avoid passing config to PebbleContainerComponent and keep knowledge of config in charm.